common/src/ethereum/signer.rs

```diff
@@ -0,0 +1,258 @@
+            .map_err(SignerMiddlewareError::SignerError)?)
+    }
+
+    /// Helper for filling a transaction's nonce using the wallet
+    async fn fill_transaction(
+        &self,
+        tx: &mut TypedTransaction,
+        block: Option<BlockId>,
+    ) -> Result<(), Self::Error> {
+        // get the `from` field's nonce if it's set, else get the signer's nonce
+        let from = if tx.from().is_some() && tx.from() != Some(&self.address()) {
+            *tx.from().unwrap()
+        } else {
+            self.address
+        };
+        tx.set_from(from);
+
+        let nonce = maybe(tx.nonce().cloned(), self.get_transaction_count(from, block)).await?;
+        tx.set_nonce(nonce);
+        self.inner()
+            .fill_transaction(tx, block)
+            .await
+            .map_err(SignerMiddlewareError::MiddlewareError)?;
+        Ok(())
+    }
+
+    /// Signs and broadcasts the transaction. The optional parameter `block` can be passed so that
+    /// gas cost and nonce calculations take it into account. For simple transactions this can be
+    /// left to `None`.
+    async fn send_transaction<T: Into<TypedTransaction> + Send + Sync>(
+        &self,
+        tx: T,
+        block: Option<BlockId>,
+    ) -> Result<PendingTransaction<'_, Self::Provider>, Self::Error> {
+        let mut tx = tx.into();
+
+        // fill any missing fields
+        self.fill_transaction(&mut tx, block).await?;
+
+        // If the from address is set and is not our signer, delegate to inner
+        if tx.from().is_some() && tx.from() != Some(&self.address()) {
+            return self
+                .inner
+                .send_transaction(tx, block)
+                .await
+                .map_err(SignerMiddlewareError::MiddlewareError);
+        }
+
+        if self.signer.is_walletconnect() && !self.legacy {
+            let tx_hash = self
+                .signer
+                .send_transaction(&tx)
+                .await
+                .map_err(SignerMiddlewareError::SignerError)?;
+            Ok(PendingTransaction::new(tx_hash, self.inner.provider()))
+        } else {
+            // if we have a nonce manager set, we should try handling the result in
+            // case there was a nonce mismatch
+            let signed_tx = self.sign_transaction(tx).await?;
+
+            // Submit the raw transaction
+            self.inner
+                .send_raw_transaction(signed_tx)
+                .await
+                .map_err(SignerMiddlewareError::MiddlewareError)
+        }
+    }
+
+    /// Signs a message with the internal signer, or if none is present it will make a call to
+    /// the connected node's `eth_call` API.
+    async fn sign<T: Into<Bytes> + Send + Sync>(
+        &self,
+        data: T,
+        _: &Address,
+    ) -> Result<Signature, Self::Error> {
+        self.signer
+            .sign_message(data.into())
+            .await
+            .map_err(SignerMiddlewareError::SignerError)
+    }
+}
```

Comment:
Why did we copy all of this over from ethers?

Reply:
I mean why do we have our own Signer type with all the same methods plus one extra and our own SignerMiddleware and SignerMiddlewareError?

Reply:
Ok, refactored it. Check [this](https://github.com/radicle-dev/radicle-cli/compare/171dad3c0b2bba1a7a257d712084e7ff83f23099..49ad2de82da8cd7a938e3448676583d6cc5d806a) please.